saas/renewals.py

```diff
@@ -242,19 +243,61 @@ def trigger_expiration_notices(at_time=None, nb_days=15, dry_run=False):
     LOGGER.info(
         "trigger notifications for subscription expiring within [%s,%s[ ...",
         lower, upper)
+    orgs_processed = []
     for subscription in Subscription.objects.valid_for(
-            auto_renew=False, ends_at__gte=lower, ends_at__lt=upper,
-            plan__auto_renew=True):
-        LOGGER.info("trigger expires soon for %s", subscription)
-        if not dry_run:
-            try:
-                signals.expires_soon.send(sender=__name__,
-                    subscription=subscription, nb_days=nb_days)
-            except Exception as err: #pylint:disable=broad-except
-                # We use `Exception` because the email server might be
-                # unavailable but ConnectionRefusedError is not a subclass
-                # of RuntimeError.
-                LOGGER.exception("error: %s", err)
+            auto_renew=False, ends_at__gte=lower, ends_at__lt=upper):
+        org = subscription.organization
+        plan = subscription.plan
+
+        if subscription.auto_renew == True:
+            if plan.renewal_type == plan.AUTO_RENEW and org.id not in orgs_processed:
```

Comment:
I see what `orgs_processed` is meant for. Best is to generate a SQL `ORDER BY organization_id` and only trigger the signal when the `organization.id` changes. Example:

    prev_organization = None
    for subscription in Subcription.order_by('organization'):
        if prev_organization and subscription.organization != prev_organization:
             ... do something only once per organization...
       prev_organization = organization
    if prev_organization and subscription.organization != prev_organization:
       ... do something only once per organization...
    
The sort is executed by the database (once and fast) and it avoids to build data structures in memory in Python, which can be a problem on very large datasets.